File: src/connectors/haveibeenpwned.js
import { BaseConnector } from './base.js';
import https from 'https';

export class HaveIBeenPwnedConnector extends BaseConnector {
    constructor(config, state, telemetry) {
        super(config, state, telemetry);
        this.name = 'HaveIBeenPwned';
    }

    async run(investigationId, phaseId, inputType, inputValue) {
        const runId = this.state.recordConnectorRun(investigationId, phaseId, this.name, inputType, inputValue);
        this.telemetry.connectorStart(investigationId, this.name, phaseId, { type: inputType, value: inputValue });

        try {
            const entities = [];

            const breaches = await this._apiGet(`/api/v3/breachedaccount/${encodeURIComponent(inputValue)}?truncateResponse=false`);

            if (Array.isArray(breaches)) {
                for (const breach of breaches) {
                    entities.push({
                        type: 'breach',
                        data: {
                            name: breach.Name,
                            title: breach.Title,
                            domain: breach.Domain,
                            date: breach.BreachDate,
                            added_date: breach.AddedDate,
                            modified_date: breach.ModifiedDate,
                            pwn_count: breach.PwnCount,
                            description: breach.Description?.replace(/<[^>]+>/g, '').slice(0, 500),
                            data_classes: breach.DataClasses,
                            is_verified: breach.IsVerified,
                            is_sensitive: breach.IsSensitive,
                            email: inputValue,
                            source: 'haveibeenpwned'
                        },
                        confidence: breach.IsVerified ? 0.95 : 0.7
                    });

                    if (breach.Domain) {
                        entities.push({
                            type: 'domain',
                            data: { name: breach.Domain, source: `breach:${breach.Name}` },
                            confidence: 0.6
                        });
                    }
                }
            }

            const pastes = await this._apiGet(`/api/v3/pasteaccount/${encodeURIComponent(inputValue)}`);
            if (Array.isArray(pastes)) {
                for (const paste of pastes) {
                    entities.push({
                        type: 'breach',
                        data: {
                            name: `Paste: ${paste.Title || paste.Id}`,
                            source_type: 'paste',
                            paste_source: paste.Source,
                            date: paste.Date,
                            email_count: paste.EmailCount,
                            email: inputValue,
                            source: 'haveibeenpwned_paste'
                        },
                        confidence: 0.6
                    });
                }
            }

            let newCount = 0;
            for (const entity of entities) {
                const added = this.state.addEntity(investigationId, entity.type, entity.data, entity.confidence || 0.5);
                if (added.new) newCount++;
                this.telemetry.entityDiscovered(investigationId, entity.type, added.new, this.name);
            }

            this.state.completeConnectorRun(runId, 'completed', newCount, null, null, 0);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'completed', entitiesFound: newCount, input: { type: inputType, value: inputValue }
            });
            return { status: 'completed', entities, newCount };
        } catch (e) {
            this.state.completeConnectorRun(runId, 'failed', 0, null, e.message, 1);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'failed', entitiesFound: 0, input: { type: inputType, value: inputValue }
            });
            return { status: 'failed', error: e.message };
        }
    }

    _apiGet(path) {
        return new Promise((resolve, reject) => {
            const opts = {
                hostname: 'haveibeenpwned.com',
                path,
                headers: {
                    'User-Agent': 'cyclops-osint',
                    'Accept': 'application/json'
                }
            };
            const apiKey = process.env.HIBP_API_KEY;
            if (apiKey) opts.headers['hibp-api-key'] = apiKey;

            https.get(opts, res => {
                if (res.statusCode === 404) return resolve([]);
                if (res.statusCode === 401) return reject(new Error('HIBP API key required (set HIBP_API_KEY)'));
                if (res.statusCode === 429) return reject(new Error('HIBP rate limited'));
                let data = '';
                res.on('data', chunk => { data += chunk; });
                res.on('end', () => {
                    try { resolve(JSON.parse(data)); } catch { resolve([]); }
                });
            }).on('error', reject);
        });
    }

    async healthCheck() {
        const key = process.env.HIBP_API_KEY;
        return { ok: !!key, version: 'v3', hasApiKey: !!key };
    }
}

File: cyclops/connectors/haveibeenpwned.py
import os
import aiohttp
from cyclops.connectors.base import BaseConnector


class HaveIBeenPwnedConnector(BaseConnector):
    def __init__(self, config, state, telemetry):
        super().__init__(config, state, telemetry)
        self.name = "HaveIBeenPwned"
        self.api_key = os.environ.get("HIBP_API_KEY")

    async def run(self, investigation_id, phase_id, input_type, input_value):
        run_id = self.state.record_connector_run(investigation_id, phase_id, self.name, input_type, input_value)
        self.telemetry.connector_start(investigation_id, self.name, phase_id, {"type": input_type, "value": input_value})

        try:
            entities = []
            headers = {"User-Agent": "CYCLOPS-OSINT", "Accept": "application/json"}
            if self.api_key:
                headers["hibp-api-key"] = self.api_key

            async with aiohttp.ClientSession(headers=headers) as session:
                url = f"https://haveibeenpwned.com/api/v3/breachedaccount/{input_value}?truncateResponse=false"
                async with session.get(url, timeout=aiohttp.ClientTimeout(total=15)) as resp:
                    if resp.status == 404:
                        pass
                    elif resp.status == 401:
                        raise Exception("HIBP API key required (set HIBP_API_KEY)")
                    elif resp.status == 429:
                        raise Exception("HIBP rate limited")
                    elif resp.status >= 400:
                        raise Exception(f"HIBP HTTP {resp.status}")
                    else:
                        breaches = await resp.json()
                        for breach in breaches:
                            entities.append({
                                "type": "breach",
                                "data": {
                                    "name": breach["Name"],
                                    "domain": breach.get("Domain"),
                                    "date": breach.get("BreachDate"),
                                    "count": breach.get("PwnCount"),
                                    "data_classes": breach.get("DataClasses", []),
                                    "address": input_value,
                                    "source": "haveibeenpwned",
                                },
                                "confidence": 0.9,
                            })
                            if breach.get("Domain"):
                                entities.append({
                                    "type": "domain",
                                    "data": {"name": breach["Domain"], "source": "hibp_breach"},
                                    "confidence": 0.6,
                                })

            new_count = 0
            for entity in entities:
                added = self.state.add_entity(investigation_id, entity["type"], entity["data"], entity.get("confidence", 0.5))
                if added["new"]:
                    new_count += 1
                self.telemetry.entity_discovered(investigation_id, entity["type"], added["new"], self.name)

            self.state.complete_connector_run(run_id, "completed", new_count)
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "completed", "entitiesFound": new_count, "input": {"type": input_type, "value": input_value}})
            return {"status": "completed", "entities": entities, "newCount": new_count}
        except Exception as e:
            self.state.complete_connector_run(run_id, "failed", 0, error=str(e))
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "failed", "input": {"type": input_type, "value": input_value}})
            return {"status": "failed", "error": str(e)}

    async def health_check(self):
        return {"ok": True, "hasApiKey": bool(self.api_key)}
